In [1]:
import pandas as pd

In [2]:
import tensorflow as tf

In [3]:
from tensorflow.keras.preprocessing.text import Tokenizer

In [4]:
from tensorflow.keras.utils import pad_sequences

In [5]:
from tensorflow.keras.utils import to_categorical

In [6]:
from tensorflow.keras import Sequential

In [7]:
from tensorflow.keras.layers import LSTM, Embedding, Bidirectional, Dense

In [64]:
import numpy as np

In [8]:
df = pd.read_csv("../../../dados/medium_data.csv")

In [9]:
df = df[['id', 'title']]
df

id                                              title
0        1  A Beginner’s Guide to Word Embedding with Gens...
1        2  Hands-on Graph Neural Networks with PyTorch & ...
2        3                       How to Use ggplot2 in Python
3        4  Databricks: How to Save Files in CSV on Your L...
4        5  A Step-by-Step Implementation of Gradient Desc...
...    ...                                                ...
6503  6504  “We” vs “I” — How Should You Talk About Yourse...
6504  6505                   How Donald Trump Markets Himself
6505  6506      Content and Marketing Beyond Mass Consumption
6506  6507  5 Questions All Copywriters Should Ask Clients...
6507  6508             How To Write a Good Business Blog Post

[6508 rows x 2 columns]

In [10]:
df['texto'] = df['title'].str.replace(u"\xa0", u' ')

In [119]:
tokenizer = Tokenizer( oov_token='OOV')

In [122]:
tokenizer.fit_on_texts(df['texto'])

In [13]:
dicionario = tokenizer.word_index

In [14]:
linha = df['texto'][0]
linha

'A Beginner’s Guide to Word Embedding with Gensim Word2Vec Model'

In [15]:
dicionario

{'to': 1,
 'the': 2,
 'strong': 3,
 'a': 4,
 'of': 5,
 'how': 6,
 'and': 7,
 'in': 8,
 'your': 9,
 'markup': 10,
 'for': 11,
 'you': 12,
 'with': 13,
 'is': 14,
 'data': 15,
 'class': 16,
 'why': 17,
 'h3': 18,
 'i': 19,
 'what': 20,
 'on': 21,
 'learning': 22,
 'from': 23,
 'an': 24,
 'be': 25,
 'my': 26,
 'writing': 27,
 'are': 28,
 'it': 29,
 'can': 30,
 'using': 31,
 'ux': 32,
 'design': 33,
 'machine': 34,
 'about': 35,
 'do': 36,
 'not': 37,
 'python': 38,
 'ai': 39,
 'life': 40,
 'that': 41,
 'when': 42,
 'should': 43,
 'we': 44,
 '5': 45,
 'science': 46,
 'make': 47,
 'need': 48,
 'time': 49,
 'as': 50,
 'more': 51,
 'at': 52,
 '3': 53,
 'business': 54,
 'or': 55,
 'have': 56,
 'work': 57,
 'new': 58,
 'don’t': 59,
 'by': 60,
 'up': 61,
 'write': 62,
 'get': 63,
 'guide': 64,
 'use': 65,
 'will': 66,
 'marketing': 67,
 'ways': 68,
 'deep': 69,
 'part': 70,
 'best': 71,
 '1': 72,
 'first': 73,
 'better': 74,
 '2019': 75,
 'product': 76,
 'analysis': 77,
 'neural': 78,
 'know': 7

In [16]:
tokenizer.texts_to_sequences([linha])

[[4, 664, 64, 1, 447, 1492, 13, 2425, 3608, 97]]

# Codigo para gerar sequencias com base na lista de palavras
[
    [4, 664],
    [4, 664, 64],
    [4, 664, 64, 1],
    [4, 664, 64, 1, 447],
    [4, 664, 64, 1, 447, 1492],
    [4, 664, 64, 1, 447, 1492, 13],
    [4, 664, 64, 1, 447, 1492, 13, 2425],
    [4, 664, 64, 1, 447, 1492, 13, 2425, 3608],
    [4, 664, 64, 1, 447, 1492, 13, 2425, 3608, 97],
]

In [17]:
lista_sequencias = []
maior_sequencia = 0
for linha in df['texto']:
    # linha = df['texto'][0]
    token_list = tokenizer.texts_to_sequences([linha])[0]
    tamanho_sequencia = len(token_list)
    for i in range(1, tamanho_sequencia):
        n_grams = token_list[:i + 1]
        lista_sequencias.append(n_grams)
        if tamanho_sequencia > maior_sequencia:
            maior_sequencia = tamanho_sequencia
# print(lista_sequencias)

In [37]:
# lista_sequencias
lista_sequencias[:10]

[[4, 664],
 [4, 664, 64],
 [4, 664, 64, 1],
 [4, 664, 64, 1, 447],
 [4, 664, 64, 1, 447, 1492],
 [4, 664, 64, 1, 447, 1492, 13],
 [4, 664, 64, 1, 447, 1492, 13, 2425],
 [4, 664, 64, 1, 447, 1492, 13, 2425, 3608],
 [4, 664, 64, 1, 447, 1492, 13, 2425, 3608, 97],
 [1827, 21]]

In [19]:
maior_sequencia

38

In [20]:
sequencias_padded = pad_sequences(lista_sequencias)

In [21]:
sequencias_padded.shape

(47594, 38)

In [41]:
sequencias_padded[4]

array([   0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    4,
        664,   64,    1,  447, 1492])

In [23]:
linha = sequencias_padded[3]
entrada = linha[:-1]
entrada

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   4, 664,  64,   1])

In [24]:
saida = linha[-1]
saida

447

In [42]:
sequencias_padded[3]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   4, 664,  64,   1, 447])

In [43]:
X[3]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   4, 664,  64,   1])

In [45]:
Y_temp[3]

447

In [46]:
Y[3]

array([0., 0., 0., ..., 0., 0., 0.])

In [25]:
X = sequencias_padded[:, :-1]
X[3]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   4, 664,  64,   1])

In [26]:
Y_temp = sequencias_padded[:, -1]
Y_temp[3]

447

In [27]:
Y = to_categorical(Y_temp)

In [28]:
Y.shape

(47594, 8543)

In [29]:
model = Sequential()
model.add(Embedding(Y.shape[1], 100, input_length=maior_sequencia))
model.add(Bidirectional(LSTM(150)))
# model.add(Dense(64, activation='relu'))
model.add(Dense(Y.shape[1], activation='softmax'))
model.summary()

C:\Users\anton\AppData\Roaming\Python\Python310\site-packages\keras\src\layers\core\embedding.py:86: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding (Embedding)                │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ bidirectional (Bidirectional)        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [30]:
adam = tf.keras.optimizers.Adam(learning_rate=0.001)

In [31]:
model.compile(loss='categorical_crossentropy', optimizer = adam, metrics = ['accuracy'])

In [32]:
model.fit(X, Y, epochs=20)

Epoch 1/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 122s 78ms/step - accuracy: 0.0568 - loss: 7.3408
Epoch 2/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 112s 75ms/step - accuracy: 0.1280 - loss: 6.2934
Epoch 3/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 107s 72ms/step - accuracy: 0.1592 - loss: 5.7267
Epoch 4/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 109s 73ms/step - accuracy: 0.1846 - loss: 5.2336
Epoch 5/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 108s 73ms/step - accuracy: 0.2119 - loss: 4.7504
Epoch 6/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 113s 76ms/step - accuracy: 0.2435 - loss: 4.3140
Epoch 7/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 128s 86ms/step - accuracy: 0.2918 - loss: 3.9037
Epoch 8/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 113s 76ms/step - accuracy: 0.3416 - loss: 3.5076
Epoch 9/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 110s 74ms/step - accuracy: 0.3988 - loss: 3.1659
Epoch 10/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 107s 72ms/step - accuracy: 0.4561 - loss: 2.8258
Epoch 11/20
1488/1488 ━━━━━━━━━━━━━━━━━━━━ 104s 70ms/step - accuracy: 0.5009 - loss: 2.54

In [51]:
meu_texto = "machine learning is"

In [53]:
meu_texto_sequences = tokenizer.texts_to_sequences([meu_texto])

In [59]:
entrada_sequences = pad_sequences(meu_texto_sequences, maxlen=maior_sequencia, padding='pre')

In [36]:
tokenizer.sequences_to_texts([[37]])

['networks']

In [61]:
saida = model.predict(entrada_sequences)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step


In [62]:
saida.shape

(1, 8543)

In [65]:
np.argmax(saida)

37

In [68]:
saida_sequences = []

In [70]:
saida_sequences.extend(meu_texto_sequences)

In [72]:
saida_sequences[0].append(37)

In [73]:
saida_sequences

[[34, 22, 14, 37]]

In [74]:
tokenizer.sequences_to_texts(saida_sequences)

['machine learning is not']

In [84]:
def complete_a_frase( texto ):
    global maior_sequencia
    meu_texto_sequences = tokenizer.texts_to_sequences(texto)
    entrada_sequences = pad_sequences(meu_texto_sequences, maxlen=maior_sequencia, padding='pre')
    saida = model.predict(entrada_sequences)
    indice_proxima_palavra = np.argmax(saida)
    saida_sequences = []
    saida_sequences.extend(meu_texto_sequences)
    saida_sequences[0].append(indice_proxima_palavra)
    return tokenizer.sequences_to_texts(saida_sequences)
    

In [116]:
complete_a_frase(["maxine learn"])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step


['learn to']

In [127]:
meu_texto_sequences = tokenizer.texts_to_sequences(["maxine learn brazil"])

In [128]:
meu_texto_sequences

[[1, 127, 4469]]

In [129]:
tokenizer.sequences_to_texts(meu_texto_sequences)

['OOV learn brazil']